python/tskit/trees.py

```diff
@@ -10387,3 +10291,134 @@ def write_ms(
                         )
             else:
                 print(file=output)
+
+def _rand_pow_range_finder(
+    operator,
+    operator_dim: int,
+    rank: int,
+    depth: int,
+    num_vectors: int,
+    rng: np.random.Generator,
+    range_sketch: np.ndarray = None,
+    ) -> np.ndarray:
+    """
+    Algorithm 9 in https://arxiv.org/pdf/2002.01387
+    """
+    assert num_vectors >= rank > 0, "num_vectors should be larger than rank"
+    if range_sketch is None:
+        test_vectors = rng.normal(size=(operator_dim, num_vectors))
+        Q = test_vectors
+    else:
+        Q = range_sketch
+    for _ in range(depth):
+        Q = np.linalg.qr(Q).Q
+        Q = operator(Q)
+    Q = np.linalg.qr(Q).Q
+    return Q[:, :rank]
+
+def _rand_svd(
+    operator,
+    operator_dim: int,
+    rank: int,
+    depth: int,
+    num_vectors: int,
+    rng: np.random.Generator,
+    range_sketch: np.ndarray = None,
+    ) -> (np.ndarray, np.ndarray, np.ndarray, float):
+    """
+    Algorithm 8 in https://arxiv.org/pdf/2002.01387
+    """
+    assert num_vectors >= rank > 0
+    Q = _rand_pow_range_finder(
+        operator, operator_dim, num_vectors, depth, num_vectors, rng, range_sketch
+    )
+    C = operator(Q).T
+    U_hat, D, V = np.linalg.svd(C, full_matrices=False)
+    U = Q @ U_hat
+
+    error_factor = np.power(
+            1 + 4 * np.sqrt(2 * operator_dim / (rank - 1)), 
+            1 / (2 * depth + 1)
+            )
+    error_bound = D[-1] * (1 + error_factor)
+    return U[:, :rank], D[:rank], V[:rank], Q, error_bound
+
+def _genetic_relatedness_vector_individual(
```

Comment:
These are best seen as methods of the TreeSequence class because they have the first argument as a tree sequence. I'd refactor as 
```
    def _genetic_relatedness_vector_individual(self, arr, indices, mode...):
          ...
    def _genetic_relatedness_vector_node(self, ...)
         ...
```

Reply:
They are now tree sequence methods. SVD functions were moved into the `pca()` function. Only the class definition is left outside.